zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -1074,11 +1125,53 @@ where
                 }
             }
             PoolType::Transparent => {
+                #[cfg(not(feature = "transparent-inputs"))]
                 return Err(Error::UnsupportedChangeType(output_pool));
             }
         }
     }
 
+    // This reserves the ephemeral addresses even if transaction construction fails.
+    // It is not worth the complexity of being able to unreserve them, because there
+    // are few failure modes after this point that would allow us to do so.
+    #[cfg(feature = "transparent-inputs")]
+    {
+        let ephemeral_outputs: Vec<(usize, &ChangeValue)> = proposal_step
+            .balance()
+            .proposed_change()
+            .iter()
+            .enumerate()
+            .filter(|(_, change_value)| matches!(change_value.output_pool(), PoolType::Transparent))
+            .collect();
+        let num_ephemeral_outputs =
+            u32::try_from(ephemeral_outputs.len()).map_err(|_| Error::ProposalNotSupported)?;
+
+        let addresses_and_metadata = wallet_db
+            .reserve_next_n_ephemeral_addresses(account_id, num_ephemeral_outputs)
```

Comment:
What is the use case for having multiple ephemeral outputs from a single step? It seems like there should only ever be a single ephemeral output; having multiple inputs for the second transaction doesn't provide any benefit I can determine.

Reply:
It was mainly for symmetry with the existing handling of change. There's no way in the protobuf schema to enforce that only one `ChangeValue` has `isEphemeral` set (although I guess there are other constraints that aren't enforced by the schema, anyway). At this point I'd prefer not to do another refactoring to only allow one ephemeral output.

Reply:
I did refactor the proposal creation side to use `Option<NonNegativeAmount>`, but left this as-is.